processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferArrayGrouper.java

```diff
@@ -0,0 +1,281 @@
+                                 .mapToInt(AggregatorFactory::getMaxIntermediateSize)
+                                 .sum();
+
+    return keySerde.keySize() +                                         // key size
+           (int) Math.ceil((double) cardinalityWithMissingValue / 8) +  // total used flags size
+           cardinalityWithMissingValue * recordSize;                    // total values size
+  }
+
+  public BufferArrayGrouper(
+      final Supplier<ByteBuffer> bufferSupplier,
+      final KeySerde<KeyType> keySerde,
+      final ColumnSelectorFactory columnSelectorFactory,
+      final AggregatorFactory[] aggregatorFactories,
+      final int cardinality
+  )
+  {
+    Preconditions.checkNotNull(aggregatorFactories, "aggregatorFactories");
+    Preconditions.checkArgument(cardinality > 0, "Cardinality must a non-zero positive number");
+
+    this.bufferSupplier = Preconditions.checkNotNull(bufferSupplier, "bufferSupplier");
+    this.keySerde = Preconditions.checkNotNull(keySerde, "keySerde");
+    this.aggregators = new BufferAggregator[aggregatorFactories.length];
+    this.aggregatorOffsets = new int[aggregatorFactories.length];
+    this.cardinalityWithMissingValue = cardinality + 1;
+
+    int offset = 0;
+    for (int i = 0; i < aggregatorFactories.length; i++) {
+      aggregators[i] = aggregatorFactories[i].factorizeBuffered(columnSelectorFactory);
+      aggregatorOffsets[i] = offset;
+      offset += aggregatorFactories[i].getMaxIntermediateSize();
+    }
+    recordSize = offset;
+  }
+
+  @Override
+  public void init()
+  {
+    if (!initialized) {
+      final ByteBuffer buffer = bufferSupplier.get().duplicate();
+
+      buffer.position(0);
+      buffer.limit(keySerde.keySize());
+      keyBuffer = buffer.slice();
+
+      final int usedBufferEnd = keySerde.keySize() + (int) Math.ceil((double) cardinalityWithMissingValue / 8);
+      buffer.position(keySerde.keySize());
+      buffer.limit(usedBufferEnd);
+      usedFlagBuffer = buffer.slice();
+
+      buffer.position(usedBufferEnd);
+      buffer.limit(buffer.capacity());
+      valBuffer = buffer.slice();
+
+      reset();
+
+      initialized = true;
+    }
+  }
+
+  @Override
+  public boolean isInitialized()
+  {
+    return initialized;
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key)
+  {
+    // BufferArrayGrouper is used only for dictionary-indexed single-value string dimensions.
+    // Here, the key contains the dictionary-encoded value of the grouping key
+    // and we use it as the index for the aggregation array.
+
+    final ByteBuffer fromKey = checkAndGetKeyBuffer(keySerde, key);
+    if (fromKey == null) {
+      // This may just trigger a spill and get ignored, which is ok. If it bubbles up to the user, the message will
+      // be correct.
+      return Groupers.DICTIONARY_FULL;
+    }
+
+    // The first index is reserved for missing value which is represented by -1.
+    final int dimIndex = fromKey.getInt() + 1;
```

Comment:
This code assumes that the key is 4 bytes. The constructor of this class should verify that `keySerde.keySize()` is 4 and throw an illegal argument exception otherwise.